### PR TITLE
cuda : keep columns in registers for cooperative softmax

### PR DESCRIPTION
**Pull request by sirohikartik**
## Overview

Resolves the TODO at line 145 in `ggml/src/ggml-cuda/softmax.cu`:
```cpp
// TODO: Template to allow keeping ncols in registers if they fit
```

In `soft_max_f32_parallelize_cols_single_row`, each row is parallelized across all SMs in the grid using cooperative groups for `ncols > 8192`. Previously, the kernel hardcoded `n_elem_per_thread = 4` in a strided loop and discarded loaded values across passes, requiring 5 global memory accesses (3 reads + 2 writes) per row.

This PR templates `soft_max_f32_parallelize_cols_single_row` and `soft_max_f32_parallelize_cols` on `int n_elem_per_thread = 0`:
- When columns fit in registers (`ncols <= grid_size * n_elem_per_thread`), values are read from global memory once, kept in registers across the max and exp/sum phases, and written to global memory once.
- Global memory traffic is reduced from 5 passes to 2 passes (60% reduction).
- Preserves fallback to the original strided loop when `ncols` exceeds the register threshold.

### Benchmark Results (NVIDIA Tesla T4 GPU)

Evaluated on an NVIDIA Tesla T4 (Compute Capability 7.5, 40 SMs = 10,240 threads total) across real-world LLM vocabulary and context sizes:

| Model / Target | `ncols` | Elements / Thread ($N$) | Baseline ($\mu\text{s}$) | Templated ($\mu\text{s}$) | Speedup | Max Diff vs CPU |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| Minimum threshold | 8,192 | 1 | 11.8 | 10.6 | **1.11x** | $1.40 \times 10^{-9}$ |
| Context 16k | 16,384 | 2 | 12.0 | 10.7 | **1.13x** | $1.46 \times 10^{-9}$ |
| LLaMA-2 Vocab | 32,000 | 4 | 12.6 | 11.2 | **1.12x** | $4.37 \times 10^{-9}$ |
| Vocab 64k | 64,000 | 8 | 14.7 | 12.2 | **1.20x** | $7.89 \times 10^{-9}$ |
| LLaMA-3 Vocab | 128,256 | 16 | 18.9 | 14.3 | **1.32x** | $1.22 \times 10^{-8}$ |
| Qwen-2 Vocab | 152,064 | 16 | 19.4 | 14.5 | **1.34x** | $1.12 \times 10^{-8}$ |
| Gemma Vocab | 256,000 | 32 | 25.7 | 15.0 | **1.71x** | $1.31 \times 10^{-8}$ |

Key observations:
- **Up to 1.71x speedup** on large vocabularies (Gemma 256k latency drops from $25.7\,\mu\text{s}$ to $15.0\,\mu\text{s}$, a 41.6% reduction).
- Latency scales much more flatly as vocabulary size grows ($+4.4\,\mu\text{s}$ vs $+13.9\,\mu\text{s}$ on baseline).
- Numerical accuracy matches double-precision CPU reference within sub-float32 epsilon ($< 1.31 \times 10^{-8}$).

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - Gemini 3.8 Flash was used to assist in writing trial kernels, debugging CuMetal barriers on macOS, and benchmarking on Tesla T4.

### CODE REVIEW

**Comment by sirohikartik** [comment]
@am17an @ggerganov @gaugarg-nv @ORippler @JohannesGaessler

**Comment by ORippler** [comment]
Please also measure E2E influence on LLM-inference of this optimization. I explicitly left it as a TODO as we hit diminishing returns due to Amdahl's law at the time of implementing this (this was pre-spec-dec).

**Comment by am17an** [comment]
it's still useful in general. Softmax shows up in all sorts of places apart from sampling.

**Comment by JohannesGaessler** [comment]
Please demonstrate any concrete use case where this is actually making a difference, otherwise I consider a review not worth the opportunity cost.

**Comment by sirohikartik** [comment]
@JohannesGaessler

I have access to an H100  and an H200 , so I can run an end-to-end comparison between master and this PR on either.

Would `llama-bench` results be sufficient, or do you specifically want a realistic generation workload?

If `llama-bench` is sufficient, are there particular models, context sizes, batch sizes, or other configurations you would like me to test? I can keep the hardware, model, and settings identical between master and the PR and report the relative change.

**Comment by am17an** [comment]
It's very unlikely this is going to have a e2e affect on LLM inference. You need to find a ggml use-case where softmax over large sizes is more common rather than 1 per forward pass.

**Comment by sirohikartik** [comment]
@am17an Do you have a particular ggml operation or workload in mind that would be a good candidate to benchmark? Otherwise I can search for the current softmax call sites and identify workloads where large ncols occurs frequently.

**Comment by am17an** [comment]
calculating cross entropy loss would be a good candidate. You can check if it helps in a backward pass as that is over `[n_vocab, n_tokens]`, so large batch sizes should benefit

**Comment by am17an** [comment]
Actually it wouldn't help there either, it doesn't hit this function

**Comment by am17an** [comment]
It would help in `-fa off` at larger contexts as there we do a softmax over `n_kv`.

**Comment by sirohikartik** [comment]
@am17an test-backend-ops.cpp already has test cases added specifically to benchmark this kernel path. It runs through ggml_backend_graph_compute, allocating real GGML CUDA tensors and measuring end-to-end execution time.

Should I go ahead and run this benchmark?

**Comment by JohannesGaessler** [comment]
That is not end-to-end performance. Please show a meaningful difference using `llama-bench`.

**Comment by kvnloo** [comment]
Thanks @sirohikartik for working on this.

One detail that may help choose the end-to-end workload: the cooperative path requires no mask/sinks and unit softmax scale, among other gates, so `-fa off` alone doesn't establish that the changed kernel runs. `llama-bench` also documents that it excludes sampling.

I'd keep this PR focused on register caching. A useful next check looks like a base-vs-PR generation comparison with backend sampling enabled, confirming that the specialized kernel actually executes and measuring total generation latency. The sampler configuration matters too: an early small top-k can remove the large-row workload we're trying to measure.

For a separate follow-up, would it make sense to reuse that workload and its correctness checks to determine whether the remaining cost is inside softmax or in the intermediate writes and launches between sampling operations? That could guide whether further kernel tuning or a small fusion experiment is worth pursuing, while preserving sampler order, numerical behavior, random-number state handling, and the existing fallback.

I wouldn't make that broader work a prerequisite for this PR; the immediate question is whether this change produces a useful workload-level improvement.

This is based on source review, not GPU measurements.
